flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/restore/AbstractRocksDBRestoreOperation.java

```diff
@@ -0,0 +1,193 @@
+	protected final KeyGroupRange keyGroupRange;
+	protected final int keyGroupPrefixBytes;
+	protected final int numberOfTransferringThreads;
+	protected final CloseableRegistry cancelStreamRegistry;
+	protected final ClassLoader userCodeClassLoader;
+	protected final ColumnFamilyOptions columnOptions;
+	protected final DBOptions dbOptions;
+	protected final Map<String, StateColumnFamilyHandle> kvStateInformation;
+	protected final File instanceBasePath;
+	protected final File instanceRocksDBPath;
+	protected final String dbPath;
+	protected List<ColumnFamilyHandle> columnFamilyHandles;
+	protected List<ColumnFamilyDescriptor> columnFamilyDescriptors;
+	protected final StateSerializerProvider<K> keySerializerProvider;
+	protected final RocksDBNativeMetricOptions nativeMetricOptions;
+	protected final MetricGroup metricGroup;
+	protected final Collection<KeyedStateHandle> restoreStateHandles;
+
+	protected RocksDB db;
+	protected ColumnFamilyHandle defaultColumnFamilyHandle;
+	protected RocksDBNativeMetricMonitor nativeMetricMonitor;
+	protected boolean isKeySerializerCompatibilityChecked;
+
+	protected AbstractRocksDBRestoreOperation(
+		KeyGroupRange keyGroupRange,
+		int keyGroupPrefixBytes,
+		int numberOfTransferringThreads,
+		CloseableRegistry cancelStreamRegistry,
+		ClassLoader userCodeClassLoader,
+		Map<String, StateColumnFamilyHandle> kvStateInformation,
+		StateSerializerProvider<K> keySerializerProvider,
+		File instanceBasePath,
+		File instanceRocksDBPath,
+		DBOptions dbOptions,
+		ColumnFamilyOptions columnOptions,
+		RocksDBNativeMetricOptions nativeMetricOptions,
+		MetricGroup metricGroup,
+		@Nonnull Collection<KeyedStateHandle> stateHandles) {
+		this.keyGroupRange = keyGroupRange;
+		this.keyGroupPrefixBytes = keyGroupPrefixBytes;
+		this.numberOfTransferringThreads = numberOfTransferringThreads;
+		this.cancelStreamRegistry = cancelStreamRegistry;
+		this.userCodeClassLoader = userCodeClassLoader;
+		this.kvStateInformation = kvStateInformation;
+		this.keySerializerProvider = keySerializerProvider;
+		this.instanceBasePath = instanceBasePath;
+		this.instanceRocksDBPath = instanceRocksDBPath;
+		this.dbPath = instanceRocksDBPath.getAbsolutePath();
+		this.dbOptions = dbOptions;
+		this.columnOptions = columnOptions;
+		this.nativeMetricOptions = nativeMetricOptions;
+		this.metricGroup = metricGroup;
+		this.restoreStateHandles = stateHandles;
+		this.columnFamilyHandles = new ArrayList<>(1);
+		this.columnFamilyDescriptors = Collections.emptyList();
+	}
+
+	public void openDB() throws IOException {
+		db = RocksDBOperationUtils.openDB(dbPath, columnFamilyDescriptors,
+			columnFamilyHandles, columnOptions, dbOptions);
+		// remove the default column family which is located at the first index
+		defaultColumnFamilyHandle = columnFamilyHandles.remove(0);
+		// init native metrics monitor if configured
+		nativeMetricMonitor = nativeMetricOptions.isEnabled() ?
```

Comment:
I think this also means that we should include `IOUtils.closeQuietly(nativeMetricMonitor)` in the close method of this class?

Reply:
Yes, also noticed this when debugging the resource leak issue, will include in next commit.